src/test/mon/test_election.cc

```diff
@@ -0,0 +1,369 @@
+{
+  Owner *o = electors[to];
+  queue_message(from, to, [o, from, e] {
+      o->receive_propose(from, e);
+    });
+}
+
+void Election::claim_victory(int from, int to, epoch_t e, const set<int>& members)
+{
+  Owner *o = electors[to];
+  queue_message(from, to, [o, from, e, members] {
+      o->receive_victory_claim(from, e, members);
+    });
+}
+
+void Election::accept_victory(int from, int to, epoch_t e)
+{
+  Owner *o = electors[to];
+  queue_message(from, to, [o, from, e] {
+      o->receive_victory_ack(from, e);
+    });
+}
+
+int Election::run_timesteps(int max)
+{
+  vector< function<void()> > current_m;
+  int steps = 0;
+  for (; (!max || steps < max) && // we have timesteps left AND ONE OF
+	 (!messages.empty() || // there are messages pending.
+	  !election_stable()); // somebody's not happy and will act in future
+       ++steps) {
+    current_m.clear();
+    current_m.swap(messages);
+    for (auto& m : current_m) {
+      m();
+    }
+    for (auto o : electors) {
+      o.second->notify_timestep();
+    }
+  }
+
+  return steps;
+}
+
+void Election::start_one(int who)
+{
+  assert(who < static_cast<int>(electors.size()));
+  electors[who]->logic.start();
+}
+
+void Election::start_all() {
+  for (auto e : electors) {
+    e.second->logic.start();
+  }
+}
+
+bool Election::election_stable()
+{
+  // see if anybody has a timer running
+  for (auto i : electors) {
+    if (i.second->timer_steps != -1)
+      return false;
+  }
+  return true;
+}
+
+void Election::block_messages(int from, int to)
+{
+  blocked_messages[from].insert(to);
+}
+void Election::block_bidirectional_messages(int a, int b)
+{
+  block_messages(a, b);
+  block_messages(b, a);
+}
+void Election::unblock_messages(int from, int to)
+{
+  blocked_messages[from].erase(to);
+}
+void Election::unblock_bidirectional_messages(int a, int b)
+{
+  unblock_messages(a, b);
+  unblock_messages(b, a);
+}
+
+
+TEST(election, binary_runs_at_all)
+{
+  cout << "hello test world!" << std::endl;
+  ASSERT_EQ(0, 0);
+}
+
+TEST(election, single_startup_election_completes)
+{
+  for (int starter = 0; starter < 5; ++starter) {
+    Election election(5);
+    election.start_one(starter);
+    // This test is not actually legit since you should start
+    // all the ElectionLogics, but it seems to work
+    int steps = election.run_timesteps(0);
+    cerr << "ran in " << steps << " timesteps" << std::endl;
+    ASSERT_TRUE(election.election_stable());
+    Owner *first_o = election.electors[0];
+    int leader = first_o->logic.get_acked_leader();
+    int epoch = first_o->logic.get_epoch();
+    for (auto i : election.electors) {
+      Owner *o = i.second;
+      ASSERT_EQ(leader, o->logic.get_acked_leader());
+      ASSERT_EQ(epoch, o->logic.get_epoch());
+    }
+  }
+}
+
+TEST(election, everybody_starts_completes)
+{
+  Election election(5);
+  election.start_all();
+  int steps = election.run_timesteps(0);
+  cerr << "ran in " << steps << " timesteps" << std::endl;
```

Comment:
i'd suggest avoid print anything (especially to stderr!) if everything is fine so far.

Reply:
Ah shoot, I changed this in a later patch but it's a bit messy to pull back for this PR. Let me see if I can make it work.

Reply:
Okay got it!